src/kv/memstore.rs

```diff
@@ -108,76 +113,147 @@ impl Write for WriteTransaction<'_> {
+        let mut ms = MemStore::new();
+        ms.put("k1", "v1".as_bytes()).await?;
 
         let rt = ms.read().await?;
-        assert_eq!(true, rt.has("bar").await?);
-        assert_eq!(Some(b"baz".to_vec()), rt.get("bar").await?);
+        assert_eq!(true, rt.has("k1").await?);
+        assert_eq!(Some(b"v1".to_vec()), rt.get("k1").await?);
 
         Ok(())
     }
 
     #[async_std::test]
-    async fn delete() -> std::result::Result<(), StoreError> {
+    async fn test_write_transaction() -> std::result::Result<(), StoreError> {
         let mut ms = MemStore::new();
-        ms.put("bar", "foo".as_bytes()).await?;
+        ms.put("k1", "v1".as_bytes()).await?;
+        ms.put("k2", "v2".as_bytes()).await?;
 
+        // Test put then commit.
         let wt = ms.write().await?;
-        assert_eq!(true, wt.has("bar").await?);
-        wt.del("bar").await?;
-        assert_eq!(false, wt.has("bar").await?);
-        wt.put("bar", b"overwrite").await?;
-        assert_eq!(true, wt.has("bar").await?);
-        assert_eq!(Some(b"overwrite".to_vec()), wt.get("bar").await?);
-        wt.del("bar").await?;
+        assert_eq!(true, wt.has("k1").await?);
+        assert_eq!(true, wt.has("k2").await?);
+        wt.put("k1", b"overwrite").await?;
         wt.commit().await?;
+        assert_eq!(Some(b"overwrite".to_vec()), ms.get("k1").await?);
+        assert_eq!(Some(b"v2".to_vec()), ms.get("k2").await?);
 
-        let rt = ms.read().await?;
-        assert_eq!(false, rt.has("bar").await?);
+        // Test put then rollback.
+        let wt = ms.write().await?;
+        wt.put("k1", b"should be rolled back").await?;
+        wt.rollback().await?;
+        assert_eq!(Some(b"overwrite".to_vec()), ms.get("k1").await?);
+
+        // Test del then commit.
+        let wt = ms.write().await?;
+        wt.del("k1").await?;
+        assert_eq!(false, wt.has("k1").await?);
+        wt.commit().await?;
+        assert_eq!(false, ms.has("k1").await?);
+
+        // Test del then rollback.
+        assert_eq!(true, ms.has("k2").await?);
+        let wt = ms.write().await?;
+        wt.del("k2").await?;
+        assert_eq!(false, wt.has("k2").await?);
+        wt.rollback().await?;
+        assert_eq!(true, ms.has("k2").await?);
+
+        // Test overwrite multiple times then commit.
+        let wt = ms.write().await?;
+        wt.put("k2", b"overwrite").await?;
+        wt.del("k2").await?;
+        wt.put("k2", b"final").await?;
+        wt.commit().await?;
+        assert_eq!(Some(b"final".to_vec()), ms.get("k2").await?);
+
+        // Test as_read.
+        let wt = ms.write().await?;
+        wt.put("k2", b"new value").await?;
+        let rt = wt.as_read();
+        assert_eq!(true, rt.has("k2").await?);
+        assert_eq!(Some(b"new value".to_vec()), rt.get("k2").await?);
 
         Ok(())
     }
+
+    #[async_std::test]
+    async fn test_isolation() {
+        let ms = MemStore::new();
```

Comment:
[minifix]

`use std::time::Duration` to avoid the repetitions below. Can be local to this function.

Reply:
the long qualifiers dont bother me, probably bc i'm still new and like to see where things come from?